### DISCUSSION TITLE
How to get the atom index from image x,y coordinates

### DISCUSSION THREAD

**Discussion by jonas87** [rdkit/rdkit#4694] [Q&A]
I'm looking to use rdkit in a small Python gui application where a rdkit drawn svg of a molecule would be shown in a window, and the user can then click on the atoms to select them. Once the index of these atoms that has been clicked on is known the program would use the highlightAtoms method and update the svg to show the selected atoms.

The part I don't really see is how to go from the x,y coordinates of the mouse click event in the svg to the index of the corresponding atom in the molecule.

In other words, let's say I have a 500x500 svg of a molecule, and the user clicked on an atom with x,y coordinates (50, 79).
How do I find the index of the atom at x,y coordinates (50, 79)?

**Comment by DavidACosgrove** [comment]
Hi Jonas,
This is possible, though without access to my computer I’m hazy as to the
details. There’s a method on the drawing object called something like
drawCoordsToAtomCoords that you pass the mouse coords to and it will return
the corresponding coordinates in the molecule coordinate frame. You can
then loop over the atoms in the molecule and find the one that is nearest.
  I’m afraid you will have to dig into the moldraw2d object to find the
exact name of the function. If you’re using Python you can use the help
method. If c++, the include file will have it.
I hope this helps. Please let me know if you have any other questions.
Dave


On Sun, 14 Nov 2021 at 19:54, jonas87 ***@***.***> wrote:

> I'm looking to use rdkit in a small Python gui application where a rdkit
> drawn svg of a molecule would be shown in a window, and the user can then
> click on the atoms to select them. Once the index of these atoms that has
> been clicked on is known the program would use the highlightAtoms method
> and update the svg to show the selected atoms.
>
> The part I don't really see is how to go from the x,y coordinates of the
> mouse click event in the svg to the index of the corresponding atom in the
> molecule.
>
> In other words, let's say I have a 500x500 svg of a molecule, and the user
> clicked on an atom with x,y coordinates (50, 79).
> How do I find the index of the atom at x,y coordinates (50, 79)?
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/rdkit/rdkit/discussions/4694>, or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACGF2FUWDDXMTN2M4QBVHXDUMAHZFANCNFSM5IAGUTKA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>
-- 
David Cosgrove
Freelance computational chemistry and chemoinformatics developer
http://cozchemix.co.uk

**Reply by jonas87** [reply]
Thanks for the pointer, I figured it out. For future reference in case someone stumbles upon this question: it's the GetDrawCoords() function of the MolDraw2D object, which takes an atom index as argument. It returns the coordinates in draw space of the specified atom, so for each mouse click event you just iterate through all atoms with calls to MolDraw2D.GetDrawCoords() and then you calculate the euclidean distance from the mouse coordinates to all the atom coordinates. The atom at the index with the shortest distance is the one that was clicked on.

**Comment by DavidACosgrove** [comment]
Hi,
Sorry for getting the logic backwards.  I’m glad you worked it out with the
flakey information. You may have done this already, or it may not be
relevant to your use case, but I find it useful to have an upper threshold
on the distance to the nearest atom so that a really bad miss with the
mouse is discarded.

Dave


On Mon, 15 Nov 2021 at 15:05, jonas87 ***@***.***> wrote:

> Thanks for the pointer, I figured it out. For future reference in case
> someone stumbles upon this question: it's the GetDrawCoords() function of
> the MolDraw2D object, which takes an atom index as argument. It returns the
> coordinates in draw space of the specified atom, so for each mouse click
> event you just iterate through all atoms with calls to
> MolDraw2D.GetDrawCoords() and then you calculate the euclidean distance
> from the mouse coordinates to all the atom coordinates. The atom at the
> index with the shortest distance is the one that was clicked on.
>
> —
> You are receiving this because you commented.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/rdkit/rdkit/discussions/4694#discussioncomment-1645261>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACGF2FSRZY4L35OHQ2M643LUMEOUZANCNFSM5IAGUTKA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>
-- 
David Cosgrove
Freelance computational chemistry and chemoinformatics developer
http://cozchemix.co.uk

**Reply by pippo1990** [reply]
cool method. Is there by any chance an equivalent of GetDrawCoords()  for a MolsToGridImage too ?


No maybe not MolsToGridImage returns a string
